### PR TITLE
[ROCm] Add Triton Backend for BF16 Grouped GEMM Backward Kernels

### PR DESCRIPTION
**Pull request by amd-wsung102**
## Summary

- Add Triton **BF16 grouped GEMM backward** kernels for ROCm, `bf16bf16bf16_grouped_grad` (dgrad / dX) and `bf16bf16bf16_grouped_wgrad` (wgrad / dW), which closes the ROCm/CUDA feature gap for grouped MoE grouped GEMM backpropagation on AMD GPUs.
- Register `mslk::bf16bf16bf16_grouped_grad` and `mslk::bf16bf16bf16_grouped_wgrad` op schemas under `USE_ROCM` in C++ and wire dispatch to the Python Triton kernels via `torch.library.impl`.
- Enable the existing `test_grouped_gemm_dgrad` / `test_grouped_gemm_wgrad` unit tests on ROCm (previously CUDA-only) and validate numerics against the fp32 reference on gfx950 (MI350X).

## Details

**Kernels** (`mslk/gemm/triton/grouped_gemm.py`):

- `_mslk_grouped_gemm_dgrad` — computes the input gradient `dX[g] = dY[g] @ W[g]` per group, producing a `[total_M, N]` output. Reuses the existing persistent grouped-GEMM structure (`program_id`-strided tile walk across the concatenated groups using the `M_sizes` prefix offsets), so it shares the same launch/scheduling mechanics as the forward `_mslk_grouped_gemm`.
- `_mslk_grouped_gemm_wgrad` — computes the weight gradient `dW[g] = dY[g].T @ X[g]` per group, producing a `[G, N, K]` output. Tiles the `[N, K]` output and reduces along the per-group M dimension. An `OUTPUT_ACCUM` `constexpr` selects between writing the result (bf16/fp16) and accumulating in-place into a pre-allocated fp32 buffer (`dW += ...`), matching the CUDA CUTLASS op's `output_accum` semantics.
- Both kernels are autotuned for AMD (CDNA) with `matrix_instr_nonkdim=16` and `waves_per_eu` tuning over block sizes / stages / warps.
- Host-side guard (`_assert_m_sizes_min`): ensures every *non-empty* group must have `m_size >= 8` (zero-token groups are allowed and skipped inside the kernel). Sub-8 groups land on an MFMA-tile edge case; callers should pad the group (round up to a multiple of 8).
- `M_sizes` is `int64`; tile counters (`iterated_tiles`) and `m_size` are kept in int64 inside the kernels for Triton type-stability across the per-group control-flow block.

**C++ changes** (`csrc/gemm/gemm_ops.cpp`):

- Register `bf16bf16bf16_grouped_grad` / `bf16bf16bf16_grouped_wgrad` schemas under `#ifdef USE_ROCM` (mirrors the existing CUDA-only schema definitions). The CUDA build continues to use the CUTLASS implementations unchanged.
- No C++ implementation/Meta registrations are added on the ROCm path. Dispatch is handled by the Python `@torch.library.impl(..., "CUDA")` registrations, which take effect when `mslk.gemm.triton.grouped_gemm` is imported.

**Wiring** (`mslk/gemm/__init__.py`):

- On HIP builds (`torch.version.hip is not None`), import `mslk.gemm.triton.grouped_gemm` so the `torch.library.impl` registrations run after the `.so` loads, making `torch.ops.mslk.bf16bf16bf16_grouped_{grad,wgrad}` available on AMD GPUs.

**Tests** (`test/gemm/gemm_test.py`):

- `supports_bf16()` now returns `True` for `gfx950` (MI350X) in addition to `gfx942` (MI300X), un-gating `BF16Tests` on CDNA4.
- `generate_random_splits()` gained a `min_group_size` parameter that produces group splits where every non-empty group is `>= min_group_size` (still allowing zero-token groups, still summing to exactly `total_M`).
- `test_grouped_gemm_dgrad` and `test_grouped_gemm_wgrad` are no longer CUDA-only and call `generate_random_splits(..., min_group_size=8)` to satisfy the host-side guard.

## Test Plan

- Build with `--build-variant rocm` on ROCm 7.2 / PyTorch 2.13 (`torch 2.13.0.dev+rocm7.2`, Triton 3.7).
- `pytest test/gemm/gemm_test.py -k "test_grouped_gemm_dgrad" -v` — **48 passed** on gfx950.
- `pytest test/gemm/gemm_test.py -k "test_grouped_gemm_wgrad and not zero_token" -v` — **96 passed** on gfx950 (covers `output_accum` True/False).
- CI on CUDA is unaffected. The ROCm schemas/registrations are `USE_ROCM`/`torch.version.hip` gated and the CUDA CUTLASS dispatch path is untouched.

## Performance

Measured on **MI350X (gfx950)**, `torch 2.13.0.dev+rocm7.2` + ROCm 7.2 (HIP 7.2.53211) + Triton 3.7, over Llama-3-8B-class FFN projection shapes (intermediate `N=14336`, hidden `K=4096`) plus square shapes. Numbers are **TFLOPS** (`2·total_M·N·K / time`, best-of-autotune, steady-state via `triton.testing.do_bench`); correctness is covered separately by the unittests. Tokens are split evenly across `G` groups (multiples of 8).

Columns:
- **dgrad** — `dX = dY @ W` (`bf16bf16bf16_grouped_grad`).
- **wgrad** — `dW = dY.T @ X`, plain bf16 output (`bf16bf16bf16_grouped_wgrad`).
- **wgrad+acc** — same, accumulating in-place into a pre-allocated fp32 buffer (`output_accum=True`).

### MI350X - G=8 token sweep (TFLOPS)

| total_M | N     | K    | dgrad | wgrad | wgrad+acc |
|--------:|------:|-----:|------:|------:|----------:|
| 1024    | 14336 | 4096 | 439   | 152   | 126       |
| 2048    | 14336 | 4096 | 567   | 226   | 185       |
| 4096    | 14336 | 4096 | 617   | 285   | 246       |
| 8192    | 14336 | 4096 | 657   | 307   | 297       |
| 16384   | 14336 | 4096 | 721   | 331   | 328       |

### MI350X - G=8 square shapes (TFLOPS)

| total_M | N    | K    | dgrad | wgrad | wgrad+acc |
|--------:|-----:|-----:|------:|------:|----------:|
| 4096    | 4096 | 4096 | 553   | 244   | 209       |
| 8192    | 8192 | 8192 | 674   | 297   | 269       |

### MI350X - G=16 token sweep (TFLOPS)

| total_M | N     | K    | dgrad | wgrad | wgrad+acc |
|--------:|------:|-----:|------:|------:|----------:|
| 8192    | 14336 | 4096 | 646   | 289   | 235       |
| 16384   | 14336 | 4096 | 649   | 309   | 292       |

### CODE REVIEW

File: csrc/gemm/gemm_ops.cpp [hunk 1]
```diff
@@ -64,6 +64,13 @@ TORCH_LIBRARY_FRAGMENT(mslk, m) {
   m.def("i8i8bf16(Tensor XQ, Tensor WQ, float scale, int split_k=1) -> Tensor");
   m.def(
       "i8i8bf16_dynamic(Tensor XQ, Tensor WQ, Tensor scale, int split_k=1) -> Tensor");
+  // BF16 grouped GEMM grad / wgrad via Triton: schemas only on ROCm;
+  // implementations are registered by mslk.gemm.triton.grouped_gemm via
+  // torch.library.impl at Python import time.
+  m.def(
+      "bf16bf16bf16_grouped_grad(Tensor X, Tensor W, Tensor M_sizes, Tensor? out=None, int? num_sms=None) -> Tensor");
```
Comment (jwfromm) [inline csrc/gemm/gemm_ops.cpp]: I think these may be duplicated in `https://github.com/meta-pytorch/MSLK/blob/main/csrc/gemm/cutlass/bf16bf16bf16_grouped_grad.cu#L885`. Now that they support AMD and cuda we can remove the registrations in bf16bf16bf16_grouped_grad.cu and bf16bf16bf16_grouped_wgrad.cu to make it cleaner.

**Review by jwfromm — APPROVED** [review]
Left one minor nitpick, overall LGTM. Just fix lint and we should be good to go.

File: csrc/gemm/gemm_ops.cpp (hunk 1 above)
Comment (amd-wsung102) [reply csrc/gemm/gemm_ops.cpp]: Thanks! I have addressed this in commit https://github.com/meta-pytorch/MSLK/pull/388/changes/0f9be3e594261ef693e1490d41f52f214f850065.

**Review by jwfromm — APPROVED** [review]
Very nice work! Thanks for the contribution.
